File: client/js-client/LeafletMap.js
var LM = (function(){

	var map, featureLayer, tileLayer, zoom, socket, boundsTimeout, boundsTimeoutTimer;
	
  function init(divName, mapOptions, tileLayerOptions, options){
    map =  L.map(divName, mapOptions);
    map.setView(options.center, options.zoom);
    featureLayer = L.layerGroup().addTo(map);
    if (tileLayerOptions )
    {
      tileLayer =  new L.tileLayer(tileLayerOptions.osmUrl, {attribution: tileLayerOptions.osmAttribution});
      tileLayer.addTo(map);
    }
    if(options.mousePositionControl)
    {
      L.control.mousePosition().addTo(map);
    }
    if (options.onMoveend)
    {
      socket = options.onMoveend;
      map.on('moveend', changeRegistration);
    }
    if (options.boundsTimeout)
    {
      boundsTimeout = options.boundsTimeout *1000;
    }
    changeRegistration();
  }

  function changeRegistration(){
      var message = {};
      message.function = "register"
      message.zoom = map.getZoom();
      message.bounds = map.getBounds();
      socket.timeQuery = new Date().getTime();
      socket.send(JSON.stringify(message));
      if (boundsTimeoutTimer) clearTimeout(boundsTimeoutTimer);
      boundsTimeoutTimer = setTimeout(changeRegistration,boundsTimeout); 
  } 
	   
  function getMap(){
    return map;
  }

  function getZoom(){
    return map.getZoom();
  }

  function addToMap(feature, animation, popupContent){
    if (typeof feature === 'undefined') return;
    if(popupContent.length > 0)
    {
      function onMouseover(e) {
        var popupOptions, latlng;
        popupOptions = {closeButton:false ,autoPan:false , minWidth: 200, maxWidth: 200, offset:new L.Point(120,-20)};
        L.popup(popupOptions).setLatLng(e.latlng).setContent(popupContent).openOn(map);
      }

      function onMouseout(e) {
        LM.getMap().closePopup();
      }      
      
      feature.on('mouseover',onMouseover);
      feature.on('mouseout', onMouseout);
    }
    featureLayer.addLayer(feature);
    if (animation == true)
    {
      feature.start();
    }
  } 
       
  function removeFeatures(vessel){
    if (typeof vessel.vector !="undefined")
    {
       featureLayer.removeLayer(vessel.vector);
    }
    if (typeof vessel.polygon !="undefined")
    {
       if (typeof vessel.polygon.stop ==='function')
       {
           vessel.polygon.stop();
       }
       featureLayer.removeLayer(vessel.polygon);
    }
    if (typeof vessel.feature !="undefined")
    {
      if (typeof vessel.feature.stop ==='function')
      {
         vessel.feature.stop();
      }
      featureLayer.removeLayer(vessel.feature);
    }
  }
  /* return puplic API */
  return {
		init: init,
		getMap: getMap,
    getZoom: getZoom,
    addToMap: addToMap,
    removeFeatures: removeFeatures
  }
})();

	

File: client/dart-client/LeafletMap.dart
library LMap;

import "dart:html";
import "dart:async";
import 'dart:json';

import 'packages/js/js.dart' as js;
import 'ais-html5.dart';
import 'MapFeature.dart';

/*--------------------------------------------------------------------------------------*/
/*-                                                                                    -*/
/*-                                class LeafletMap                           -*/
/*-                                                                                    -*/
/*--------------------------------------------------------------------------------------*/

class LeafletMap {
  js.Proxy map;
  js.Proxy featureLayerGroup;
  var boundsTimeout;
  var boundsTimeoutTimer;
  
  LeafletMap(String elementid, js.Proxy mapOptions, js.Proxy initOptions, js.Proxy tileLayerOptions){
    boundsTimeout = initOptions['boundsTimeout']*1000;
    js.scoped(() {
        map = new js.Proxy(js.context.L.Map, elementid, mapOptions);
        featureLayerGroup = new js.Proxy(js.context.L.LayerGroup);
        map.addLayer(featureLayerGroup);
        js.retain(featureLayerGroup);
        var osm = new js.Proxy(js.context.L.TileLayer,tileLayerOptions['tileURL'], tileLayerOptions);
        map.addLayer(osm);
        var mouseOptions = initOptions['mousePosition'];
        if( mouseOptions != false)
        {
          var mousePosition = new js.Proxy(js.context.L.Control.MousePosition, mouseOptions);
          mousePosition.addTo(map);
        }
        map.setView(new js.Proxy(js.context.L.LatLng, initOptions['lat'], initOptions['lon']),initOptions['zoom']);
        js.retain(map);
        map.on('moveend', new js.Callback.many(moveendHandler));
      });
     changeRegistration();
  }
  
   moveendHandler(e)
   {
     changeRegistration();
   }
  
  changeRegistration(){
    int zoom = getZoom();
    var boundsArray = getBounds().split(",");
    var _southWest = {"lng":double.parse(boundsArray[0]),"lat":double.parse(boundsArray[1])};
    var _northEast= {"lng":double.parse(boundsArray[2]),"lat":double.parse(boundsArray[3])};
    var bounds = {"_southWest": _southWest,"_northEast":_northEast};

    Map message = new Map();
    message['function'] = 'register';
    message['zoom'] =  zoom;
    message['bounds'] = bounds;
    socket.send(stringify(message));
    
    boundsTimeoutTimer = new Timer(new Duration(milliseconds:boundsTimeout),changeRegistration);  
  }
  
  String getBounds(){
    String bBox;
    js.scoped((){
      bBox = map.getBounds().toBBoxString();
    });
    return bBox;
  }

  int getZoom() {
    int zoom;
    js.scoped((){
      zoom = map.getZoom();
    });
    return zoom;
  }

  void setView(num latitude, num longitude, int zoom) {
    var _pos;
    js.scoped(() {
      _pos = new js.Proxy(js.context.L.LatLng, latitude, longitude);
      map.setView(_pos,12);
    });
  }

  removeFeatures(vessel){
    if (vessel.vector != null)
    {
      vessel.vector.remove();
    }
    if (vessel.polygon != null)
    {
      if (vessel.polygon.animated == true)
      {
        vessel.polygon.stopAnimation();
      }
      vessel.polygon.remove();
    }
    if (vessel.feature != null)
    {
      if (vessel.feature.animated == true)
      {
        vessel.feature.stopAnimation();
      }
      vessel.feature.remove();
    }
  }

  closePopup(){
    js.scoped((){
      map.closePopup();
    });
  }

  openPopup(Popup popup){
    js.scoped((){
      map.openPopup(popup._popup);
    });
  }
}

/*--------------------------------------------------------------------------------------*/
/*-                                                                                    -*/
/*-                        class Coord                                                 -*/
/*-                                                                                    -*/
/*--------------------------------------------------------------------------------------*/

class Coord {
  final num latitude;
  final num longitude;
  const Coord(this.latitude, this.longitude);
}

Coord getCoord() {
  js.scoped(() {
    var latlng = mapFeature.getLatLng();
    return new Coord(latlng.latitude, latlng.longitude);
  });
}

void setCoord(Coord coord) {
  js.scoped(() {
    var latlng = new js.Proxy(js.context.L.LatLng, coord.latitude, coord.longitude);
    mapFeature.setLatLng(latlng);
    mapFeature.update();
  });
}
/*--------------------------------------------------------------------------------------*/
/*-                                                                                    -*/
/*-                                class Popup                                        -*/
/*-                                                                                    -*/
/*--------------------------------------------------------------------------------------*/

class Popup{
  js.Proxy _popup;

  Popup( js.Proxy latlng, String content, Map options) {
    js.scoped(() {
      var popupOptions = options;
      var offsetPoint = new js.Proxy(js.context.L.Point, popupOptions['offset'][0],popupOptions['offset'][1]);
      popupOptions['offset'] = offsetPoint;
      popupOptions = js.map(popupOptions);
      _popup= new js.Proxy(js.context.L.Popup, popupOptions);
      _popup.setLatLng(latlng);
      _popup.setContent(content);
      js.retain(_popup);
    });
  }
}

